File: arm-modbus/src/main/java/io/github/iur/arm/modbus/param/UdpParam.java
package io.github.iur.arm.modbus.param;

import com.serotonin.modbus4j.ModbusMaster;
import com.serotonin.modbus4j.ip.IpParameters;
import com.serotonin.modbus4j.ip.udp.UdpMaster;

import io.github.iur.arm.modbus.ModbusParam;

/**
 * UDP参数
 */
public class UdpParam implements ModbusParam<UdpParam> {

    private final IpParameters mParameters;
    /**
     * 超时
     */
    private int timeout = 500;
    /**
     * 重试
     */
    private int retries = 2;
    /**
     * 是否验证响应中的从站ID
     */
    private boolean validateResponse;

    private UdpParam(String host, int port) {
        mParameters = new IpParameters();
        mParameters.setHost(host);
        mParameters.setPort(port);
    }

    public static UdpParam create(String host, int port) {
        return new UdpParam(host, port);
    }

    public String getHost() {
        return mParameters.getHost();
    }

    public UdpParam setHost(String host) {
        mParameters.setHost(host);
        return this;
    }

    public int getPort() {
        return mParameters.getPort();
    }

    public UdpParam setPort(int port) {
        mParameters.setPort(port);
        return this;
    }

    public boolean isEncapsulated() {
        return mParameters.isEncapsulated();
    }

    public UdpParam setEncapsulated(boolean encapsulated) {
        mParameters.setEncapsulated(encapsulated);
        return this;
    }

    public int getTimeout() {
        return timeout;
    }

    public UdpParam setTimeout(int timeout) {
        this.timeout = timeout;
        return this;
    }

    public int getRetries() {
        return retries;
    }

    public UdpParam setRetries(int retries) {
        this.retries = retries;
        return this;
    }

    public boolean isValidateResponse() {
        return validateResponse;
    }

    public UdpParam setValidateResponse(boolean validateResponse) {
        this.validateResponse = validateResponse;
        return this;
    }

    @Override
    public ModbusMaster createModbusMaster() {
        UdpMaster master = new UdpMaster(mParameters, isValidateResponse());
        master.setRetries(getRetries());
        master.setTimeout(getTimeout());
        return master;
    }
}

File: arm-modbus/src/main/java/io/github/iur/arm/modbus/param/UdpParam.kt
package io.github.iur.arm.modbus.param

import com.serotonin.modbus4j.ModbusMaster
import com.serotonin.modbus4j.ip.IpParameters
import io.github.iur.arm.modbus.AppUdpMaster
import io.github.iur.arm.modbus.ModbusParam

/**
 * UDP 参数
 */
class UdpParam private constructor(
    host: String,
    port: Int,
) : ModbusParam {
    private val ipParameters: IpParameters =
        IpParameters().apply {
            this.host = host
            this.port = port
        }

    override var timeout: Int = ModbusParam.DEFAULT_TIMEOUT
    override var retries: Int = ModbusParam.DEFAULT_RETRIES

    /**
     * 是否验证响应中的从站 ID
     */
    var validateResponse: Boolean = false

    var host: String
        get() = ipParameters.host
        set(value) {
            ipParameters.host = value
        }

    var port: Int
        get() = ipParameters.port
        set(value) {
            ipParameters.port = value
        }

    var encapsulated: Boolean
        get() = ipParameters.isEncapsulated
        set(value) {
            ipParameters.setEncapsulated(value)
        }

    override fun createModbusMaster(): ModbusMaster {
        val master = AppUdpMaster(ipParameters, validateResponse, port)
        master.retries = retries
        master.timeout = timeout
        return master
    }

    companion object {
        @JvmStatic
        fun create(
            host: String,
            port: Int,
        ): UdpParam = UdpParam(host, port)
    }
}
